clients/include/utility.hpp

```diff
@@ -968,6 +969,269 @@ inline void host_csr_to_csc(int                     M,
+        int end   = csr_row_ptr[std::min(M, block_dim * i + block_dim)] - csr_base;
+
+        std::vector<int> temp(nb, 0);
+
+        for(int j = start; j < end; j++)
+        {
+            int blockCol = (csr_col_ind[j] - csr_base) / block_dim;
+            temp[blockCol]         = 1;
+        }
+
+        for(int j = 0; j < nb; j++)
+        {
+            if(temp[j] == 1)
+            {
+                bsr_col_ind[colIndex] = j + bsr_base;
+                colIndex++;
+            }
+        }
+    }
+
+    // find bsr values array
+    for(int i = 0; i < M; i++)
+    {
+        int blockRow = i / block_dim;
+
+        int start = csr_row_ptr[i] - csr_base;
+        int end   = csr_row_ptr[i + 1] - csr_base;
+
+        for(int j = start; j < end; j++)
+        {
+            int blockCol = (csr_col_ind[j] - csr_base) / block_dim;
+
+            colIndex = -1;
+            for(int k = bsr_row_ptr[blockRow] - bsr_base;
+                k < bsr_row_ptr[blockRow + 1] - bsr_base;
+                k++)
+            {
+                if(bsr_col_ind[k] - bsr_base == blockCol)
+                {
+                    colIndex = k - (bsr_row_ptr[blockRow] - bsr_base);
+                    break;
+                }
+            }
+
+            assert(colIndex != -1);
+
+            int blockIndex = 0;
+            if(direction == HIPSPARSE_DIRECTION_ROW)
+            {
+                blockIndex = (csr_col_ind[j] - csr_base) % block_dim + (i % block_dim) * block_dim;
+            }
+            else
+            {
+                blockIndex
+                    = ((csr_col_ind[j] - csr_base) % block_dim) * block_dim + (i % block_dim);
+            }
+
+            int index = (bsr_row_ptr[blockRow] - bsr_base) * block_dim * block_dim
+                                  + colIndex * block_dim * block_dim + blockIndex;
+
+            bsr_val[index] = csr_val[j];
+        }
+    }
+}
+
+template <typename T>
+inline void host_bsr_to_csr(hipsparseDirection_t    direction,
+                            int                     Mb,
+                            int                     Nb,
+                            int                     block_dim,
+                            hipsparseIndexBase_t    bsr_base,
+                            const std::vector<int>& bsr_row_ptr,
+                            const std::vector<int>& bsr_col_ind,
+                            const std::vector<T>&   bsr_val,
+                            hipsparseIndexBase_t    csr_base,
+                            std::vector<int>&       csr_row_ptr,
+                            std::vector<int>&       csr_col_ind,
+                            std::vector<T>&         csr_val)
+{
+    int m    = Mb * block_dim;
+    int n    = Nb * block_dim;
+    int nnzb = bsr_row_ptr[Mb] - bsr_row_ptr[0];
+
+    csr_row_ptr.resize(m + 1, 0);
+    csr_col_ind.resize(nnzb * block_dim * block_dim, 0);
+    csr_val.resize(nnzb * block_dim * block_dim, static_cast<T>(0));
```

Comment:
need to change `static_cast<T>(0)` with `make_DataType<T>(0)` else nvcc will fail